### PR TITLE
Use m2e MojoParameter related code

### LINKED ISSUE #4: Content assist & hover for complex <configuration> element
From https://github.com/AObuchow/lsp4xml-extensions-maven/issues/85

m2e has it.
For some plugin configuration elements (like maven-dependency-plugin/copy/artifactItems), m2e is able to complete the children of the element.
We should copy the responsible m2e code into the Lemminx extension

Difficulty: **..
Value (frequency*time saved): ****

### PR DESCRIPTION
**Pull request by AObuchow**
Fix #4

Signed-off-by: Andrew Obuchowicz <aobuchow@redhat.com>

### CODE REVIEW

**Comment by AObuchow** [comment]
This is currently a work in progress. No actual completion logic has been added yet, although a MojoParameter can be successfully created, which contains nested/complex plugin configuration elements.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenCompletionParticipant.java [hunk 1]
```diff
@@ -99,13 +101,18 @@ public NodeWithChildrenInsertionStrategy(String elementName) {
...
-	public MavenCompletionParticipant(MavenProjectCache cache, LocalRepositorySearcher localRepositorySearcher, RemoteRepositoryIndexSearcher indexSearcher, RepositorySystemSession repoSession, MavenPluginManager pluginManager) {
+	public MavenCompletionParticipant(MavenProjectCache cache, LocalRepositorySearcher localRepositorySearcher, RemoteRepositoryIndexSearcher indexSearcher, RepositorySystemSession repoSession, MavenSession mavenSession, MavenPluginManager pluginManager, BuildPluginManager buildPluginManager) {
 		this.cache = cache;
 		this.localRepositorySearcher = localRepositorySearcher;
 		this.indexSearcher = indexSearcher;
 		this.repoSession = repoSession;
+		this.mavenSession = mavenSession;
 		this.pluginManager = pluginManager;
+		this.buildPluginManager = buildPluginManager;
+
```
Comment (AObuchow) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenCompletionParticipant.java:115]: remove extra whitespace

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenDiagnosticParticipant.java [hunk 2]
```diff
@@ -38,11 +40,15 @@
 	private MavenProjectCache projectCache;
 	MavenPluginManager pluginManager;
 	private final RepositorySystemSession repoSession;
+	private final MavenSession mavenSession;
+	private final BuildPluginManager buildPluginManager;
 
-	public MavenDiagnosticParticipant(MavenProjectCache projectCache, MavenPluginManager pluginManager, RepositorySystemSession repoSession) {
+	public MavenDiagnosticParticipant(MavenProjectCache projectCache, MavenPluginManager pluginManager, RepositorySystemSession repoSession, MavenSession mavenSession, BuildPluginManager buildPluginManager) {
```
Comment (AObuchow) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenDiagnosticParticipant.java:46]: @mickaelistria the constructor of these "participant" classes are getting kind of long, maybe we should wrap all these objects into a container class? It'll get instantiated in `MavenPlugin` and be passed to all the participants.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenPlugin.java [hunk 3]
```diff
@@ -111,6 +119,9 @@ public void initialize(Object initializationOptions) {
 			mavenRequest = initMavenRequest(container, initializationOptions);
 			DefaultRepositorySystemSessionFactory repositorySessionFactory = container.lookup(DefaultRepositorySystemSessionFactory.class);
 			repositorySystemSession = repositorySessionFactory.newRepositorySession(mavenRequest);
+			MavenExecutionResult mavenResult = new DefaultMavenExecutionResult();
+			// TODO: MavenSession is deprecated. Investigate for alternative
```
Comment (AObuchow) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenPlugin.java:123]: Make a bug for this before merging.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenPluginUtils.java [hunk 4]
```diff
@@ -47,9 +49,11 @@ public static MarkupContent getMarkupDescription(Parameter parameter) {
 						+ "Expression: " + parameter.getExpression() + LINE_BREAK + "Default Value: " + parameter.getDefaultValue()
 						+ LINE_BREAK + parameter.getDescription());
 	}
+
```
Comment (AObuchow) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenPluginUtils.java:52]: Remove extra whitespace

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenPluginUtils.java [hunk 5]
```diff
@@ -47,9 +49,11 @@ public static MarkupContent getMarkupDescription(Parameter parameter) {
 						+ "Expression: " + parameter.getExpression() + LINE_BREAK + "Default Value: " + parameter.getDefaultValue()
 						+ LINE_BREAK + parameter.getDescription());
 	}
+	
+
 
 	public static List<Parameter> collectPluginConfigurationParameters(IPositionRequest request,
```
Comment (AObuchow) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenPluginUtils.java:55]: @mickaelistria Maybe this function should now be modified to return a `List<MojoParameter>`? Alternatively, I could make a `collectPluginConfigurationMojoParameters` method?

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PlexusConfigHelper.java [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+			MavenSession mavenSession, BuildPluginManager buildPluginManager) {
+		Class<?> clazz;
+		try {
+			clazz = mojo.getImplementationClass();
+			if (clazz == null) {
+				if (descriptor.getClassRealm() == null) {
+					// from
+					// org.eclipse.m2e.editor.mojo.MojoParameterMetadataProvider.getPluginDescriptor(ArtifactKey,
+					// IMavenExecutionContext, IProgressMonitor)
+					// TODO: Maybe this should occur in MavenPlugin.java?
```
Comment (AObuchow) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PlexusConfigHelper.java:403]: This is the only real difference in this class from m2e

Comment (AObuchow) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PlexusConfigHelper.java:403]: This TODO isn't possible as the MavenPlugin class has no PluginDescriptor. However, MavenPluginUtils does.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenDiagnosticParticipant.java (hunk 2 above)
Comment (mickaelistria) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenDiagnosticParticipant.java:46]: Isn't the mavenSession expected to contain a reference to the RepositorySystemSession or others?

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenPluginUtils.java [hunk 7]
```diff
@@ -65,8 +67,17 @@ public static MarkupContent getMarkupDescription(Parameter parameter) {
...
 					.collect(Collectors.toList());
 		}
+		// Wrap these into MojoParameters that could be used? Might be duplicating work if they're present in the MojoParameters that get retrieved afterwards
 		List<Parameter> parameters = mojosToConsiderList.stream().flatMap(mojo -> mojo.getParameters().stream())
 				.collect(Collectors.toList());
+		MavenProject project =  cache.getLastSuccessfulMavenProject(request.getXMLDocument());
+		// TODO: project could be null..
+		mavenSession.setProjects(Collections.singletonList(project));
+		// TODO: This isin't currently being used, just here for debugging purposes
+		List<MojoParameter> mojoParams = mojosToConsiderList.stream().flatMap(mojo -> {
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenPluginUtils.java:77]: you can most likely omit ` { return ` and `}` there as it's a single value expression.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MojoParameter.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Red Hat Inc. and others.
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MojoParameter.java:2]: You need to keep the copyright header of the copied file.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MojoParameter.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+ * Contributors:
+ *      Anton Tanasenko. - initial API and implementation (in m2e)
+ *******************************************************************************/
+package org.eclipse.lemminx.maven;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+// Copied from m2e
```
Comment (mickaelistria) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MojoParameter.java:17]: I'd put a note about it in the copyright header "Contributors" section, adding `Andrw Obuchowicz - copy from m2e to lemminx-maven`

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MojoParameter.java (hunk 8 above)
Comment (AObuchow) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MojoParameter.java:2]: Thanks, wasn't aware of the proper procedure.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MojoParameter.java (hunk 9 above)
Comment (AObuchow) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MojoParameter.java:17]: Sounds good

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenDiagnosticParticipant.java (hunk 2 above)
Comment (AObuchow) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenDiagnosticParticipant.java:46]: It probably is! I'll have to verify this.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenHoverParticipant.java [hunk 10]
```diff
@@ -182,19 +205,59 @@ private String collectGoal(IPositionRequest request) {
...
+				for (MojoParameter parameter : nestedParameters) {
+					if (node.getLocalName().equals(parameter.getName())) {
+						return parameter.getDescription() == null
+								? MavenPluginUtils.getMarkupDescriptionUsingParent(parameter, parentParameter)
+										.getValue()
+								: MavenPluginUtils.getMarkupDescription(parameter).getValue();
+					}
+				}
+				// Fallback case is to return parent's hover info
+				return MavenPluginUtils.getMarkupDescription(parentParameter).getValue();
```
Comment (AObuchow) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenHoverParticipant.java:260]: It’s probably best just to return null here instead, as the configuration element could be invalid

**Comment by AObuchow** [comment]
Stack overflow on the CI... not good :p

**Comment by mickaelistria** [comment]
I can reproduce the test failure and will try to investigate.

**Comment by mickaelistria** [comment]
I've added a commit on top of this patches and submitted it as https://github.com/eclipse/lemminx-maven/pull/86 to slightly improve tests and add aether-connector-okhttp as dependency (this is the actual fix)

**Comment by mickaelistria** [comment]
Merged as https://github.com/eclipse/lemminx-maven/commit/862dd9bdbd53e7b4303c867f1b09fc4e89de8f9a . This can be closed.

**Pull request closed without merging**